### PR TITLE
HPCC-31535 ALA declare KQL timespan

### PR DESCRIPTION
**Pull request by rpastrana**
- Declares request level timespan to minimize query scope
- Adds method to serialize time range as ISO 8601 Time intervals
- Supply query time interval as KQL timespan
- Replace api.loganalytics.io endpoint with api.loganalytics.azure.com. due to upcoming incompatibility

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by jakesmith** [comment]
@rpastrana - assume typo in commit comment 'tieme'

**Comment by jakesmith** [comment]
>This change is a bug fix (non-breaking change which fixes an issue).

@rpastrana - can you update the PR commit to describe what the symptoms of the bug was, and how it fixes it?

**Review by jakesmith — COMMENTED** [review]
@rpastrana - changes look sensible afaics.
Please update PR commit message to describe what bug/bug symptoms were seen and how this fixes it.

**Comment by rpastrana** [comment]
Per HPCC-31535, ALA rest queries have been timing out after 3 minutes.
Per recommendations from Daniel Polk from Azure, the in-query timestamp is applied late in the query filtering. The in-request timespan should be declared in order to minimize the scope of the query. Also, "api.loganalytics.io endpoint is being replaced by api.loganalytics.azure.com. Even though api.loganalytics.io is fully supported at this time, it is recommended to transition to the new endpoint for long-term compatibility"

**Comment by rpastrana** [comment]
@jakesmith updated.
@ghalliday please merge

File: system/logaccess/Azure/LogAnalytics/CurlClient/AzureLogAnalyticsCurlClient.cpp [hunk 1]
```diff
@@ -583,6 +590,23 @@ void AzureLogAnalyticsCurlClient::searchMetaData(StringBuffer & search, const Lo
...
+
+    char fromTimeStr[40];
+    std::strftime(fromTimeStr, sizeof(fromTimeStr), "%Y-%m-%dT%H:%M:%S", std::gmtime(&from));
+    queryTimeSpan.set(fromTimeStr);
+
+    if (to != -1)
+    {
+        char toTimeStr[40];
+        std::strftime(toTimeStr, sizeof(toTimeStr), "%Y-%m-%dT%H:%M:%S", std::gmtime(&to));
+        queryTimeSpan.appendf("/%s", fromTimeStr);
```
[the remark this answers is not in the document]
Comment (ghalliday) [reply system/logaccess/Azure/LogAnalytics/CurlClient/AzureLogAnalyticsCurlClient.cpp:606]: I think CDateTime::getString() may do what you need.

**Review by ghalliday — CHANGES_REQUESTED** [review]

File: system/logaccess/Azure/LogAnalytics/CurlClient/AzureLogAnalyticsCurlClient.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (rpastrana) [reply system/logaccess/Azure/LogAnalytics/CurlClient/AzureLogAnalyticsCurlClient.cpp]: Good copy/paste error catch. fixing

**Comment by rpastrana** [comment]
@jakesmith this change is urgent and should be made available in current rc if possible
@rwagner42 fyi

File: system/logaccess/Azure/LogAnalytics/CurlClient/AzureLogAnalyticsCurlClient.cpp [hunk 2]
```diff
@@ -978,7 +1002,10 @@ bool AzureLogAnalyticsCurlClient::fetchLog(LogQueryResultDetails & resultDetails
     populateKQLQueryString(queryString, queryIndex, options);
 
     std::string readBuffer;
-    submitKQLQuery(readBuffer, token.str(), queryString.str(), m_logAnalyticsWorkspaceID.str());
+    StringBuffer queryTimeSpan;
+    const LogAccessTimeRange & trange = options.getTimeRange();
+    azureLogAnalyticsQueryTimeSpanString(queryTimeSpan, trange.getStartt().getSimple(),trange.getEndt().isNull() ? -1 : trange.getEndt().getSimple());
```
Comment (jakesmith) [inline system/logaccess/Azure/LogAnalytics/CurlClient/AzureLogAnalyticsCurlClient.cpp:1007]: trivial/formatting: inconsistent spacing (missing space after 2nd comma)

File: system/logaccess/Azure/LogAnalytics/CurlClient/AzureLogAnalyticsCurlClient.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (jakesmith) [reply system/logaccess/Azure/LogAnalytics/CurlClient/AzureLogAnalyticsCurlClient.cpp]: minor: per above comments, did you consider using CDateTime::getString ?

**Review by jakesmith — COMMENTED** [review]
@rpastrana - 2 minor comments, but looks good.

Please respond to comments, but go ahead and squash

File: system/logaccess/Azure/LogAnalytics/CurlClient/AzureLogAnalyticsCurlClient.cpp (hunk 2 above)
Comment (rpastrana) [reply system/logaccess/Azure/LogAnalytics/CurlClient/AzureLogAnalyticsCurlClient.cpp]: missing space added.

File: system/logaccess/Azure/LogAnalytics/CurlClient/AzureLogAnalyticsCurlClient.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (rpastrana) [reply system/logaccess/Azure/LogAnalytics/CurlClient/AzureLogAnalyticsCurlClient.cpp]: Considered, chose not make changes at this point.

**Review by jakesmith — APPROVED** [review]
@rpastrana looks good.

**Review by ghalliday — APPROVED** [review]

**Pull request merged**